pandas/core/generic.py

```diff
@@ -8420,7 +8420,12 @@ def first(self: FrameOrSeries, offset) -> FrameOrSeries:
             return self
 
         offset = to_offset(offset)
-        end_date = end = self.index[0] + offset
+        if not isinstance(offset, Tick) and offset.is_on_offset(self.index[0]):
+            # GH#29623 if first value is end of period, remove offset with n = 1
+            #  before adding the real offset
+            end_date = end = self.index[0] - to_offset(offset.name) + offset
```

Comment:
`offset.base` instead of `to_offset(offset.name)`?

Reply:
Better:) thx